src/tune-hyperparameters/src/main/scala/TuneHyperparameters.scala

```diff
@@ -0,0 +1,200 @@
+  /**
+    * Private function taken from spark - waits for the task to complete
+    */
+  private def awaitResult[T](awaitable: Awaitable[T], atMost: Duration): T = {
+    try {
+      val awaitPermission = null.asInstanceOf[scala.concurrent.CanAwait]
+      awaitable.result(atMost)(awaitPermission)
+    } catch {
+      case NonFatal(t) if !t.isInstanceOf[TimeoutException] =>
+        throw new SparkException("Exception thrown in awaitResult: ", t)
+    }
+  }
+
+  /** Finds the best model
+    *
+    * @param dataset The input dataset to train.
+    * @return The trained classification model.
+    */
+  override def fit(dataset: Dataset[_]): TuneHyperparametersModel = {
+    val sparkSession = dataset.sparkSession
+    val splits = MLUtils.kFold(dataset.toDF.rdd, getNumFolds, getSeed)
+    val hyperParams = getHyperParams
+    val schema = dataset.schema
+    val executionContext = getExecutionContext
+    val (evaluationMetricColumnName, operator): (String, Ordering[Double]) =
+      EvaluationUtils.getMetricWithOperator(getModels.head, getEvaluationMetric)
+    val paramsPerRun = ListBuffer[ParamMap]()
+    for (_ <- 0 until getNumRuns) {
+      // Generate the new parameters, stepping through estimators sequentially
+      paramsPerRun +=
+        ParamMap(hyperParams.map {
+          paramMap => {
+            paramMap._2.asInstanceOf[HyperParam[_]].getParamPair(paramMap._1)
+          }
+        }.toSeq: _*)
+    }
+    val numModels = getModels.length
+
+    val metrics = splits.zipWithIndex.map { case ((training, validation), splitIndex) =>
+      val trainingDataset = sparkSession.createDataFrame(training, schema).cache()
+      val validationDataset = sparkSession.createDataFrame(validation, schema).cache()
+
+      val modelFutures = ListBuffer[Future[Model[_]]]()
+      val modelFuturesParams = ListBuffer[ParamMap]()
+      for (n <- 0 until getNumRuns) {
+        val params = paramsPerRun(n)
+        modelFuturesParams += params
+        modelFutures += Future[Model[_]] {
+          val model = getModels(n % numModels).fit(trainingDataset, params)
+          model.asInstanceOf[Model[_]]
+        }
+      }
+
+      // Unpersist training data only when all models have trained
+      Future.sequence[Model[_], Iterable](modelFutures)(implicitly, executionContext)
+        .onComplete { _ => trainingDataset.unpersist() } (executionContext)
+
+      val evaluator = new ComputeModelStatistics()
+      evaluator.set(evaluator.evaluationMetric, getEvaluationMetric)
+
+      // Evaluate the models
+      val foldMetricFutures = modelFutures.zip(modelFuturesParams).map { case (modelFuture, paramMap) =>
+        modelFuture.map { model =>
+          val scoredDataset = model.transform(validationDataset, paramMap)
+          val metrics = evaluator.transform(scoredDataset)
+          val metric = metrics.select(evaluationMetricColumnName).first()(0).toString.toDouble
+          logDebug(s"Got metric $metric for model trained with $paramMap.")
```

Comment:
We should just use the cross validator from spark with a custom set of randomly sampled params

https://spark.apache.org/docs/2.2.0/ml-tuning.html

Reply:
the cross validator does not use our model stats transform, which is faster, pipelineable and supports more metrics (and has better logging).  It also does not allow the user to specify parallelism in the 2.1 release, doesn't allow the user to specify multiple estimators (only a single one) and doesn't take in hyperparams.  I also plan to add more tricks to choosing the next parameter which wouldn't be possible if we called spark's crossvalidator directly.

Reply:
We can also add more termination criteria besides numRuns in the future, which won't be possible if we use their cross validator